### PR TITLE
feat(apm): Use Jaro–Winkler distance when comparing span descriptions for matching spans

### PR DESCRIPTION
**Pull request by dashed**
When matching two spans for the transaction comparison page, we use the Jaro–Winkler distance function to compute the string similarity. See: https://en.wikipedia.org/wiki/Jaro%E2%80%93Winkler_distance

This library is used for the Jaro edit distance function https://github.com/winkjs/wink-jaro-distance/blob/master/src/wink-jaro-distance.js

The Jaro edit function is used to implement the Jaro–Winkler function using this reference implementation: https://winkjs.org/wink-distance/string-jaro-winkler.js.html

Jaro–Winkler function returns a score between 0 and 1. The lower the Jaro–Winkler distance for two strings is, the more similar the strings are. 

Jaro–Winkler is used instead of the Levenshtein distance function because Jaro–Winkler matches strings more significantly at the beginning of the strings (and hence output lower scores). The Levenshtein distance does not have this bias.

**Before:**

screencapture-dev-getsentry-net-8000-organizations-sentry-performance-compare-internal-d724ec238b624d5ba1e9e72c5cf4e375-internal-e851957963544457819895221438ff12-2020-08-13-17_58_17

**After:**

screencapture-dev-getsentry-net-8000-organizations-sentry-performance-compare-internal-d724ec238b624d5ba1e9e72c5cf4e375-internal-e851957963544457819895221438ff12-2020-08-13-17_56_35

### CODE REVIEW

File: src/sentry/static/sentry/app/views/performance/compare/utils.tsx [hunk 1]
```diff
@@ -318,17 +323,62 @@ function createChildPairs({
...
+function jaroSimilarity(thisString: string, otherString: string): number {
+  // based on https://winkjs.org/wink-distance/string-jaro-winkler.js.html
+  // and https://en.wikipedia.org/wiki/Jaro%E2%80%93Winkler_distance
+  if (thisString === otherString) {
+    return 1;
+  }
+
+  let jaroDistance: number = jaro(thisString, otherString).distance;
+
+  const scalingFactor = 0.1;
```
Comment (AbhiPrasad) [inline src/sentry/static/sentry/app/views/performance/compare/utils.tsx:335]: Can we add a comment about this? I’d imagine we would want to change this in the future, so it’s nice to have context. Same with boost context below.

File: src/sentry/static/sentry/app/views/performance/compare/utils.tsx [hunk 2]
```diff
@@ -318,17 +323,62 @@ function createChildPairs({
...
-  const descriptionsEqual = baselineSpan.description === regressionSpan.description;
 
-  return opNamesEqual && descriptionsEqual;
+  if (!opNamesEqual) {
+    return 0;
+  }
+
+  // remove whitespace and convert string to lower case as the individual characters
+  // adds noise to the edit distance function
+  const baselineDescription = (baselineSpan.description || '')
```
Comment (AbhiPrasad) [inline src/sentry/static/sentry/app/views/performance/compare/utils.tsx:372]: Any other characters we need to consider here?

File: src/sentry/static/sentry/app/views/performance/compare/utils.tsx [hunk 3]
```diff
@@ -318,17 +323,62 @@ function createChildPairs({
...
+    .replace(/\s+/g, '')
+    .toLowerCase();
+  const regressionDescription = (regressionSpan.description || '')
+    .replace(/\s+/g, '')
+    .toLowerCase();
+
+  const score = jaroSimilarity(baselineDescription, regressionDescription);
+
+  return score >= 0.8 ? score : 0;
```
Comment (AbhiPrasad) [inline src/sentry/static/sentry/app/views/performance/compare/utils.tsx:381]: Let’s pull this up into a constant as well? I can see that is can be something we change.

**Review by AbhiPrasad — COMMENTED** [review]
Nice!

File: src/sentry/static/sentry/app/views/performance/compare/utils.tsx (hunk 2 above)
Comment (dashed) [reply src/sentry/static/sentry/app/views/performance/compare/utils.tsx:372]: What characters do you have in mind?

File: src/sentry/static/sentry/app/views/performance/compare/utils.tsx (hunk 3 above)
Comment (dashed) [reply src/sentry/static/sentry/app/views/performance/compare/utils.tsx]: Fixed in https://github.com/getsentry/sentry/pull/20224/commits/8a9ed42d953f4463c778a75edb4ef215f79a00ff

File: src/sentry/static/sentry/app/views/performance/compare/utils.tsx (hunk 1 above)
Comment (dashed) [reply src/sentry/static/sentry/app/views/performance/compare/utils.tsx:340]: Fixed in https://github.com/getsentry/sentry/pull/20224/commits/4005be21e1333ff9c22408a57261da8ecf7c7e55 

This is my best understanding of these parameters.

File: src/sentry/static/sentry/app/views/performance/compare/utils.tsx (hunk 2 above)
Comment (AbhiPrasad) [reply src/sentry/static/sentry/app/views/performance/compare/utils.tsx:381]: Good question. I’m not quite sure lol. I think it’s important to keep this mind though. As we do more testing

**Review by AbhiPrasad — APPROVED** [review]
Alright I think we can ship this and starting testing with it.

File: src/sentry/static/sentry/app/views/performance/compare/utils.tsx (hunk 2 above)
Comment (dashed) [reply src/sentry/static/sentry/app/views/performance/compare/utils.tsx:381]: This should be a good start, the Jaro–Winkler  function should be able to do most of the heavy lifting 😄

**Review by rhcarvalho — COMMENTED** [review]
Would you want to add some test cases here to document what kinds of strings we want to be considered a match and what should not be a match?

Do we match spans solely by op and description, or do we also look at the tree-structure of the transaction?

**Comment by dashed** [comment]
> Would you want to add some test cases here to document what kinds of strings we want to be considered a match and what should not be a match?

I'm planning on doing this on a follow up PR.

> Do we match spans solely by op and description, or do we also look at the tree-structure of the transaction?

Both. 

If the span's parents were "matched", then their direct children are matched. The exception to this are the root spans; which are matched as follows.

Matching any two spans (root span or otherwise) entails matching by their `op` names (strict equality) and their `description` props (the Jaro-Winkler heuristic added by this PR; previously it was strict equality).

If this match passes, then the matching procedure continues to their subtrees. Otherwise, their subtrees are independent.

**Pull request merged**